_CS141/type-level-programming.md

```diff
@@ -157,16 +157,92 @@ stwo :: SNat ('Succ ('Succ 'Zero))
 stwo :: SSucc sone
 ```
 
-## Proxy Types
+## Reification
 
-We don’t have types at runtime due to **type erasure**, so we can’t write a function that does something based on a type. 
+> Reification is the process of **converting types to values**, typically using **type classes**. This resolves to the correct instances at compile-time so the right implementation of the function is evaluated at runtime.
 
-> Proxy types allow us work around Haskell’s limitations and convert a representation of a value at the **type level** into a **term level** value. 
+There are a few ways to implement the instances for the type classes and one way (which is used in the lectures) is to use **proxy types**.
+
+### Proxy Types
+
+We don’t have types at runtime due to **type erasure**, so we can’t write a function that takes types as arguments in Haskell, even though we sometimes want to.
+
+> Proxy types provide a partial work around Haskell’s limitations **for cases where** knowing what type is given to a function as its “argument” at compile-time is sufficient.
+>
+> **TLDR.** A proxy type is a type of kind `k -> *` (where `k` can be specialised to the kind of whatever types you want to give to a function as argument). Only types of kind `*` have values, so we apply our proxy **type constructor** to some argument of kind `k` to get a type of kind `*` which then makes a suitable parameter for a function.
+
+In the following **example**, we will show how this allows us to convert a representation of a value at the **type level** into a **term level** value (reification).
 
 ```haskell
 data NatProxy (a :: Nat) = MkProxy
 ```
 
-Here, the kind of `NatProxy` is actually `Nat -> *`, meaning it takes some type `a` of kind `Nat` and gives back a type of kind `*` which has a value at run-time (kind `Nat` does not have a value at run-time).
+Here, the `NatProxy` **type constructor** is of kind `Nat -> *`, meaning it takes some type `a` of kind `Nat` and gives back a type of kind `*` which has a value at run-time (kind `Nat` does not have a value at run-time).`MkProxy` is the **data constructor** for `NatProxy` and is of type `NatProxy n`.
+
+```haskell
+class FromNat (n :: Nat) where
+  fromNat :: NatProxy n -> Int
+
+instance FromNat 'Zero where -- 'Zero is a type defined by Nat in GADT section
+  -- fromNat :: NatProxy 'Zero -> Int
+  fromNat _ = 0
+```
+
+Now with our **proxy type**, we can define a **type class** with instances that we can define the `fromNat` function for. The instance of this type class for `'Zero` is trivial.
```

Comment:
It feels like this should be above the code snippet, not below.

Reply:
```suggestion
Now with our **proxy type**, we can define a `fromNat` function with a slightly modified type. Since `NatProxy :: Nat -> *`, `NatProxy n` is a valid parameter type for a function since it is of kind `*`. We use a type class to overload this `fromNat` function so that we can provide different implementations for it, depending on which type `n` is. For example, if `n` is `'Zero` we just define `fromNat` to return `0`.
```